### PR TITLE
fix : UI changes

### PR DESCRIPTION
**Pull request by Harshit933**
## Why do we need this?

- Currently we don't have any information if the payment we done through core lightning is successful or not
- We don't have a send button on the `pay_view.dart` to send the invoice without `copy-pasting`.

## Changes done :

- Added a `error decoder` class that decodes the messages and then shows them on the screen.
- Added the submit button to pay an invoice without `copy-pasting` the `bolt11`.

### CODE REVIEW

File: lib/views/pay/pay_view.dart [hunk 1]
```diff
@@ -27,10 +32,36 @@ class _PayViewState extends State<PayView> {
...
+        GlobalComponent.showAppDialog(
+            context: context,
+            title: 'Payment failed',
+            message: decoder.message,
+            closeMsg: 'Ok',
+            imageProvided: const AssetImage('assets/images/exclamation.png'));
+      }
+    }
+    if (response != null) {
+      if (context.mounted) {
```
Comment (vincenzopalazzo) [inline lib/views/pay/pay_view.dart:56]: two question

1, Why do we always need this `contex.mounted`?
2. If really needed why do not put it at the beginning of the function if it is not mounted you return and nothing happens?

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

File: lib/views/pay/pay_view.dart (hunk 1 above)
Comment (Harshit933) [reply lib/views/pay/pay_view.dart:56]: The function call to the async function in the above part which is `await widget.provider.get<AppApi>().payInvoice(invoice: boltString);`. Thus, if we do not apply this, we get this warning.
`Don't use 'BuildContext's across async gaps.`

Comment (vincenzopalazzo) [reply lib/views/pay/pay_view.dart:56]: Right I forget that dart it is painful!

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

File: lib/views/pay/pay_view.dart [hunk 2]
```diff
@@ -27,10 +32,36 @@ class _PayViewState extends State<PayView> {
...
-        await widget.provider.get<AppApi>().payInvoice(invoice: boltString);
-    return response;
+  Future<void> payInvoice(String boltString) async {
+    AppPayInvoice? response;
+    try {
+      response =
+          await widget.provider.get<AppApi>().payInvoice(invoice: boltString);
+    } catch (e) {
+      late ErrorDecoder decoder;
+      var jsonString = e.toString().substring(e.toString().indexOf('{'));
```
[the remark this answers is not in the document]
Comment (Harshit933) [reply lib/views/pay/pay_view.dart:42]: `var jsonString = e.toString().substring(e.toString().indexOf('{'));` cuts the string from 

```
Exception: {"error":{"code":-32602,"message":"This payment is destined for ourselves. Self-payments are not supported"},"id":"(null)commando:pay#198","jsonrpc":"2.0"}
```
to

```
{"error":{"code":-32602,"message":"This payment is destined for ourselves. Self-payments are not supported"},"id":"(null)commando:pay#198","jsonrpc":"2.0"}. 
```
In this way I am able to get the json.

Basically it cuts the `Exception :` part by which I am able to get it into a map

File: lib/views/pay/pay_view.dart (hunk 1 above)
Comment (Harshit933) [reply lib/views/pay/pay_view.dart:56]: Each time after `async` function is called.

File: lib/views/pay/pay_view.dart [hunk 3]
```diff
@@ -27,10 +32,36 @@ class _PayViewState extends State<PayView> {
...
+      if (e is Exception || e is LNClientException) {
```
[the remark this answers is not in the document]
Comment (Harshit933) [reply lib/views/pay/pay_view.dart:43]: Oh yes. We can remove the if condition.

File: lib/views/pay/pay_view.dart [hunk 4]
```diff
@@ -27,10 +30,34 @@ class _PayViewState extends State<PayView> {
...
+    if (response != null) {
+      if (context.mounted) {
+        GlobalComponent.showAppDialog(
+            context: context,
+            title: 'Payment Successful',
+            message: 'Payment successfully sent',
+            closeMsg: 'Ok',
+            imageProvided: const AssetImage('assets/images/Checkmark.png'));
+      }
+    }
```
Comment (vincenzopalazzo) [inline lib/views/pay/pay_view.dart:60]:
```suggestion
```

File: lib/views/pay/pay_view.dart [hunk 5]
```diff
@@ -27,10 +30,34 @@ class _PayViewState extends State<PayView> {
...
-  Future<AppPayInvoice> payInvoice(String boltString) async {
-    final response =
-        await widget.provider.get<AppApi>().payInvoice(invoice: boltString);
-    return response;
+  Future<void> payInvoice(String boltString) async {
+    AppPayInvoice? response;
+    try {
+      response =
+          await widget.provider.get<AppApi>().payInvoice(invoice: boltString);
+    } catch (e) {
```
Comment (vincenzopalazzo) [inline lib/views/pay/pay_view.dart:38]:
```suggestion
          if (context.mounted) {
        GlobalComponent.showAppDialog(
            context: context,
            title: 'Payment Successful',
            message: 'Payment successfully sent',
            closeMsg: 'Ok',
            imageProvided: const AssetImage('assets/images/Checkmark.png'));
      }
    }
    } catch (e) {
```

File: lib/views/pay/pay_view.dart [hunk 6]
```diff
@@ -27,10 +30,34 @@ class _PayViewState extends State<PayView> {
   String? createdTime;
   String? expirationTime;
 
-  Future<AppPayInvoice> payInvoice(String boltString) async {
-    final response =
-        await widget.provider.get<AppApi>().payInvoice(invoice: boltString);
-    return response;
+  Future<void> payInvoice(String boltString) async {
+    AppPayInvoice? response;
```
Comment (vincenzopalazzo) [inline lib/views/pay/pay_view.dart:34]:
```suggestion

```

File: lib/views/pay/pay_view.dart [hunk 7]
```diff
@@ -27,10 +30,34 @@ class _PayViewState extends State<PayView> {
...
+    try {
+      response =
```
Comment (vincenzopalazzo) [inline lib/views/pay/pay_view.dart:36]:
```suggestion
      var response =
```

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

File: lib/views/pay/pay_view.dart (hunk 7 above)
Comment (Harshit933) [reply lib/views/pay/pay_view.dart:36]: I think we can remove this whole from `var response = await ....` to `await ....`

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]

**Review by vincenzopalazzo — APPROVED** [review]

**Comment by vincenzopalazzo** [comment]
Thanks!

**Pull request merged**